cmds/exp/tsort/tsort.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2012-2024 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// tsort writes to standard output a totally ordered list of items consistent
+// with a partial ordering of items contained in the input. The standard input
+// will be used if no file is specified.
+//
+// The input is a sequence of pairs of items, separated by <blank> characters.
+// Pairs of different items (e.g., "a b") indicate ordering. Pairs of identical
+// items (e.g., "c c") indicate presence, but not ordering.
+//
+// Synopsis:
+//
+//	tsort [FILE]
+//
+// Example:
+//	tsort <<EOF
+//	a b c c d e
+//	g g
+//	f g e f
+//	h h
+//	EOF
+//
+// produces an output like:
+//
+//	a
+//	b
+//	c
+//	d
+//	e
+//	f
+//	g
+//	h
+//
+// which is one valid total ordering, but this is not guaranteed, it could
+// equally be:
+//
+//	h
+//	a
+//	c
+//	d
+//	b
+//	e
+//	f
+//	g
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+func run(args []string, stdin io.Reader, stdout, stderr io.Writer) int {
```

Comment:
Thx for PR!

Could you change order for this arguments to be more consistent with the rest of codebase:
```go
func run(stdin io.Reader, stdout, stderr io.Writer, args []string) error
```
and add something like this in main:
```go
	if err := run(os.Stdin, os.Stdout, os.Stderr, args...); err != nil {
		log.Fatalf("tsort: %v", err)
	}
```

And return an error instead of integer, I think it will be better and maybe you can eliminate os.Stderr if you don't plan to write warnings.

Reply:
> Thx for PR!

@binjip978 yw! And thanks for reviewing it. :)

> Could you change order for this arguments to be more consistent with the rest of codebase:
> 
> ```go
> func run(stdin io.Reader, stdout, stderr io.Writer, args []string) error
> ```

Sure! When I next have time, I'll do that.

> and add something like this in main:
> 
> ```go
> 	if err := run(os.Stdin, os.Stdout, os.Stderr, args...); err != nil {
> 		log.Fatalf("tsort: %v", err)
> 	}
> ```
> 
> And return an error instead of integer, I think it will be better and maybe you can eliminate os.Stderr if you don't plan to write warnings.

If we used `log.Fatalf` like you suggest, it would print a timestamp before the message, which existing tsorts don't seem to do. So `2024/05/11 17:15:16 tsort: odd data count` vs `tsort: odd data count`. Would you like me to go ahead with this `log.Fatalf` pattern anyway?

Also, I do plan on writing warnings, because when existing tsorts encounter cycles in the input, they print the cycles to stderr and print the rest of the nodes in topological order to stdout.

Do you have any preferences between these options?

**Option A**
```
a
b
2024/05/11 17:15:16 tsort: cycle in data
2024/05/11 17:15:16 tsort: c
2024/05/11 17:15:16 tsort: d
e
```

**Option B**
```
a
b
2024/05/11 17:15:16 tsort: cycle in data
tsort: c
tsort: d
e
```

**Option C**
```
a
b
tsort: cycle in data
tsort: c
tsort: d
e
```

If option A or B, I'll need to find a way to use `log.Errorf` in `run` or something to make this work.

Reply:
Warnings are fine and you can just write to stderr and continue no logging prefix required. 
But log.Fatal will basically do a print and call os.Exit(1), so I think it can be only the last message. In this case I think it's ok for fatal error with logging message, most of our command doing exactly that. 

But my point mostly to return an error instead of integers, it can be non fatal error, so in main you can do os.Exit(1).

Something like this:
```go
if errors.Is(err, errNonFatal) {
  os.Exit(1)
}
```

Reply:
Okay, understood! I've pushed some changes which should address this, please take a look.